Source/Details/ASCollectionViewLayout.m

```diff
@@ -0,0 +1,162 @@
+//
+//  ASCollectionViewLayout.m
+//  AsyncDisplayKit
+//
+//  Created by Huy Nguyen on 28/2/17.
+//  Copyright © 2017 Facebook. All rights reserved.
+//
+
+#import <AsyncDisplayKit/ASCollectionViewLayout.h>
+
+#import <AsyncDisplayKit/ASCellNode+Internal.h>
+
+#import <AsyncDisplayKit/ASAssert.h>
+#import <AsyncDisplayKit/ASCollectionElement.h>
+#import <AsyncDisplayKit/ASCollectionView.h>
+#import <AsyncDisplayKit/ASCollectionViewFlowLayoutCalculator.h>
+#import <AsyncDisplayKit/ASElementMap.h>
+#import <AsyncDisplayKit/ASLayout.h>
+
+@interface ASCollectionViewLayout () {
+  ASElementMap * __weak _elementMap;
+  CGSize _contentSize;
+  NSMapTable<ASCollectionElement *, UICollectionViewLayoutAttributes *> *_elementToLayoutAttributesMap;
+  BOOL _collectionViewIsPendingLayoutUpdate;
+}
+
+@end
+
+@implementation ASCollectionViewLayout
+
+- (instancetype)initWithLayoutCalculator:(id<ASCollectionViewLayoutCalculating>)calculator
+{
+  self = [super init];
+  if (self) {
+    _calculator = calculator;
+    _elementToLayoutAttributesMap = [NSMapTable mapTableWithKeyOptions:(NSMapTableStrongMemory | NSMapTableObjectPointerPersonality) valueOptions:NSMapTableCopyIn];
+  }
+  return self;
+}
+
+- (instancetype)initWithCoder:(NSCoder *)aDecoder
+{
+  return [self initWithLayoutCalculator:[[ASCollectionViewFlowLayoutCalculator alloc] init]];
+}
+
+- (instancetype)init
+{
+  return [self initWithLayoutCalculator:[[ASCollectionViewFlowLayoutCalculator alloc] init]];
+}
+
+#pragma mark - ASCollectionViewLayoutCalculatorProviding
+
+- (id<ASCollectionViewLayoutCalculating>)collectionViewLayoutCalculator
+{
+  return _calculator;
+}
+
+- (CGSize)sizeForCalculatingCollectionViewLayout
+{
+  ASDisplayNodeAssertMainThread();
+  return self.collectionView.bounds.size;
+}
+
+#pragma mark - ASCollectionViewLayoutCalculatorResultConsuming
+
+- (void)setLayout:(ASLayout *)layout calculatedBy:(id<ASCollectionViewLayoutCalculating>)calculator forElementMap:(ASElementMap *)map
+{
+  ASDisplayNodeAssertMainThread();
+  ASDisplayNodeAssertTrue(calculator == _calculator);
+  [self invalidateLayout];
+  _collectionViewIsPendingLayoutUpdate = YES;
+  _elementMap = map;
+  [self _applyLayout:layout];
+}
+
+#pragma mark - UICollectionViewLayout overrides
+
+- (void)prepareLayout
+{
+  if (_collectionViewIsPendingLayoutUpdate == NO && _elementMap != nil) {
+    ASLayout *layout = [_calculator layoutThatFits:self.collectionView.bounds.size forElementMap:_elementMap];
+    [self _applyLayout:layout];
+  }
+}
+
+- (void)invalidateLayout
+{
+  [super invalidateLayout];
+  if (_collectionViewIsPendingLayoutUpdate == NO) {
+    _contentSize = CGSizeZero;
```

Comment:
Is this "safe" to do, in other words can we expect that UICollectionView won't call the contentSize method below until the layout is re-done?  I assume so, but just am building my own understanding of this detail.  May be worth a comment as this is a file that a reasonable number of framework users might take a peek at to try to understand.